code/datums/map_zones.dm

```diff
@@ -0,0 +1,626 @@
+	return ..()
+
+/datum/virtual_level/Destroy()
+	for(var/dir in crosslinked)
+		if(crosslinked[dir]) //Because it could be linking with itself
+			unlink(dir)
+	var/datum/space_level/level = SSmapping.z_list[z_value]
+	level.virtual_levels -= src
+	SSmapping.virtual_z_translation -= "[id]"
+	parent_map_zone.remove_virtual_level(src)
+	return ..()
+
+/datum/virtual_level/proc/mark_turfs()
+	for(var/turf/turf as anything in get_block())
+		turf.virtual_z = id
+
+/datum/virtual_level/proc/clear_reservation()
+	/// Create a dummy reservation to safeguard this space from being allocated mid-clearing in case the virtual level does get deleted
+	var/datum/dummy_space_reservation/safeguard = new(low_x, low_y, high_x, high_y, z_value)
+
+	var/area/space_area = GLOB.areas_by_type[/area/space]
+	for(var/turf/turf as anything in get_block())
+		//Reset turf
+		turf.empty(RESERVED_TURF_TYPE, RESERVED_TURF_TYPE, null, TRUE)
+		// Reset area
+		var/area/old_area = get_area(turf)
+		space_area.contents += turf
+		turf.change_area(old_area, space_area)
+		CHECK_TICK
+
+	qdel(safeguard)
+
+/datum/virtual_level/proc/get_trait(trait)
+	return traits[trait]
+
+/datum/virtual_level/proc/get_unreserved_bottom_left_turf()
+	return locate(low_x + reserved_margin, low_y + reserved_margin, z_value)
+
+/datum/virtual_level/proc/get_unreserved_top_right_turf()
+	return locate(high_x - reserved_margin, high_y - reserved_margin, z_value)
+
+/// Gets a random turf on a side of a virtual level
+/datum/virtual_level/proc/get_side_turf(dir, padding = 0, middle = FALSE)
+	var/r_low_x = low_x + reserved_margin
+	var/r_low_y = low_y + reserved_margin
+	var/r_high_x = high_x - reserved_margin
+	var/r_high_y = high_y - reserved_margin
+	if(!dir)
+		dir = pick(GLOB.cardinals)
+	var/turf/found_turf
+	if(middle)
+		switch(dir)
+			if(NORTH)
+				found_turf = locate(round((low_x + high_x) / 2), r_high_y, z_value)
+			if(SOUTH)
+				found_turf = locate(round((low_x + high_x) / 2), r_low_y, z_value)
+			if(EAST)
+				found_turf = locate(r_high_x, round((low_y + high_y) / 2), z_value)
+			if(WEST)
+				found_turf = locate(r_low_x, round((low_y + high_y) / 2), z_value)
+	else
+		switch(dir)
+			if(NORTH)
+				found_turf = locate(rand(r_low_x + padding, r_high_x - padding), r_high_y, z_value)
+			if(SOUTH)
+				found_turf = locate(rand(r_low_x + padding, r_high_x - padding), r_low_y, z_value)
+			if(EAST)
+				found_turf = locate(r_high_x, rand(r_low_y + padding, r_high_y - padding), z_value)
+			if(WEST)
+				found_turf = locate(r_low_x, rand(r_low_y + padding, r_high_y - padding), z_value)
+	return found_turf
+
+/datum/virtual_level/proc/reserve(x1, y1, x2, y2, passed_z)
+	low_x = x1
+	low_y = y1
+	high_x = x2
+	high_y = y2
+	z_value = passed_z
+	parent_level = SSmapping.z_list[z_value]
+	parent_level.virtual_levels += src
+	x_distance = high_x - low_x + 1
+	y_distance = high_y - low_y + 1
+	mark_turfs()
+
+/datum/virtual_level/proc/is_in_bounds(atom/Atom)
+	if(Atom.virtual_z() == z_value)
```

Comment:
```suggestion
	if(Atom.virtual_z() == id)
```

Reply:
If this isn't the right fix, please feel free to revert it, but I'm taking a wild guess that it'll fix the bug where people get deleted.